rx-java/rxgrpc-test/src/test/java/com/salesforce/rxgrpc/BackpressureIntegrationTest.java

```diff
@@ -18,66 +16,65 @@
+    private static final int NUMBER_OF_STREAM_ELEMENTS = 200;
+
+    private static AtomicLong clientLastValueTime;
+    private static AtomicLong serverLastValueTime;
+    private static AtomicLong clientNbOfWaits;
+    private static AtomicLong serverNumberOfWaits;
+
     private static Server server;
     private static ManagedChannel channel;
 
-    private static final int madMultipleCutoff = 100;
-    private static BackpressureDetector serverRecBPDetector = new BackpressureDetector(madMultipleCutoff);
-    private static BackpressureDetector serverRespBPDetector = new BackpressureDetector(madMultipleCutoff);
-
     @BeforeClass
     public static void setupServer() throws Exception {
         RxNumbersGrpc.NumbersImplBase svc = new RxNumbersGrpc.NumbersImplBase() {
             @Override
             public Single<NumberProto.Number> requestPressure(Flowable<NumberProto.Number> request) {
                 return request
                         .map(proto -> proto.getNumber(0))
-                        .doOnNext(i -> {
-                            serverRecBPDetector.tick();
-                            System.out.println("    --> " + i);
-                            try { Thread.sleep(50); } catch (InterruptedException e) {}
-                        })
+                        .doOnNext(i -> System.out.println("    --> " + i))
+                        .doOnNext(i -> waitIfValuesAreEqual(i, 3))
                         .last(-1)
                         .map(BackpressureIntegrationTest::protoNum);
             }
 
             @Override
             public Flowable<NumberProto.Number> responsePressure(Single<Empty> request) {
                 return Flowable
-                        .fromIterable(new Sequence(200, serverRespBPDetector))
+                        .fromIterable(IntStream.range(0, NUMBER_OF_STREAM_ELEMENTS)::iterator)
                         .doOnNext(i -> System.out.println("   <-- " + i))
+                        .doOnNext(i -> updateNumberOfWaits(serverLastValueTime, serverNumberOfWaits))
                         .map(BackpressureIntegrationTest::protoNum);
             }
 
             @Override
             public Flowable<NumberProto.Number> twoWayPressure(Flowable<NumberProto.Number> request) {
                 request
                     .map(proto -> proto.getNumber(0))
-                    .subscribe(
-                        n -> {
-                            serverRecBPDetector.tick();
-                            System.out.println("   --> " + n);
-                            try { Thread.sleep(50); } catch (InterruptedException e) {}
-                        },
-                        Throwable::printStackTrace,
-                        () -> System.out.println("Server done.")
-                    );
-
+                        .doOnNext(n -> System.out.println("   --> " + n))
+                        .doOnNext(n -> waitIfValuesAreEqual(n, 3))
+                        .subscribe();
```

Comment:
We shoukd hava a way to ensure that this Flowable has completed before doing the asserts. Any idea on how to do it ?

Reply:
With a global CountDownLatch ?